### ISSUE TITLE
TypeError: Cannot read properties of undefined (reading 'warnIfFlagDeprecated')

### ISSUE THREAD

**Issue by endreymarcell-wise** [oclif/core#528]
**Describe the bug**
I am running oclif as a dependency of [twilio-cli](https://www.npmjs.com/package/twilio-cli) (specifically, [twilio-cli-core](https://github.com/twilio/twilio-cli-core/blob/e8f132bd2b02717f28545337db91b582ca18575d/package.json#L45)). When trying to run commands using the `twilio` command line tool, I get the following error:

```
/github/home/.twilio-cli/node_modules/@oclif/core/lib/command.js:153
        this.warnIfFlagDeprecated(results.flags ?? {});
             ^

TypeError: Cannot read properties of undefined (reading 'warnIfFlagDeprecated')
    at parse (/github/home/.twilio-cli/node_modules/@oclif/core/lib/command.js:153:14)
    at async /github/home/.twilio-cli/node_modules/@twilio-labs/plugin-flex/dist/utils/parser.js:113:42
    at async FlexPluginsDeploy.init (/github/home/.twilio-cli/node_modules/@twilio-labs/plugin-flex/dist/commands/flex/plugins/deploy.js:65:[24](https://github.com/transferwise/plugin-twilio-flex/actions/runs/3272804474/jobs/5384283979#step:10:25))
    at async FlexPluginsDeploy._run (/github/home/.twilio-cli/node_modules/@oclif/core/lib/command.js:79:13)
    at async Config.runCommand (/usr/local/lib/node_modules/twilio-cli/node_modules/@oclif/config/lib/config.js:173:24)
    at async Main.run (/usr/local/lib/node_modules/twilio-cli/node_modules/@oclif/command/lib/main.js:[27](https://github.com/transferwise/plugin-twilio-flex/actions/runs/3272804474/jobs/5384283979#step:10:28):9)
    at async Main._run (/usr/local/lib/node_modules/twilio-cli/node_modules/@oclif/command/lib/command.js:43:20)
```

The exact context of the exception is this:
```javascript
    warnIfFlagDeprecated(flags) {
        for (const flag of Object.keys(flags)) {
            const deprecated = this.ctor.flags[flag]?.deprecated;
            if (deprecated) {
                this.warn((0, util_2.formatFlagDeprecationWarning)(flag, deprecated));
            }
        }
    }
    warnIfCommandDeprecated() {
        if (this.ctor.state === 'deprecated') {
            const cmdName = (0, index_1.toConfiguredId)(this.ctor.id, this.config);
            this.warn((0, util_2.formatCommandDeprecationWarning)(cmdName, this.ctor.deprecationOptions));
        }
    }
    async parse(options, argv = this.argv) {
        if (!options)
            options = this.constructor;
        const opts = { context: this, ...options };
        // the spread operator doesn't work with getters so we have to manually add it here
        opts.flags = options?.flags;
        opts.args = options?.args;
        const results = await Parser.parse(argv, opts);
        this.warnIfFlagDeprecated(results.flags ?? {});  // <-- ERROR
        return results;
    }

```

This looks like it could have been introduced by https://github.com/oclif/core/pull/527 ([1.19.0](https://github.com/oclif/core/releases/tag/1.19.0))
Our job was still passing 21 hours ago; the PR linked was merged 16 hours ago; our first job today was 3 hours ago and it already failed.

My computer still @oclif/core v1.16.7 locally, and it seems to work without problems.

**To Reproduce**
1. `npm install -g twilio-cli`
2. `twilio plugins:install @twilio-labs/plugin-flex`
3. `twilio profiles:create`

I understand it would be more useful to provide a direct reproduction which only relies on oclif, but if there's indeed an error in the code, it can hopefully be identified by the stack trace.

**Expected behavior**
Commands run successfully.

**Environment (please complete the following information):**
 - Ubuntu docker container with Node 16 installed, running on Github actions

**Comment by bmcgrogan** [comment]
Yeah also facing the same issue. Build pipeline was successful in building using twilio-cli 3.4.1 yesterday. Tried the same build again today and got this error. Tried upgrading both flex-plugin and twilio-cli versions, but no success. Seems something has broken upstream, perhaps.

**Comment by moretalk** [comment]
Same issue here with twilio-cli 5.1.0

**Comment by endreymarcell-wise** [comment]
Hey @mdonnalley, thanks for picking this up! I'm not sure what would be special about my codebase, and as you can see, others are also reporting the same issue.
Nevertheless, I'm happy to provide more details as much as I can (codebase is my company's, not mine).

**Comment by shrutiburman** [comment]
@mdonnalley, this might help: the issue somehow is prominent in Windows machines and Mac Silicon chips. I somehow couldn't reporduce it in my darwin os.

**Comment by moretalk** [comment]
I am reproducing it on Ubuntu via Github Actions.

**Comment by mdonnalley** [comment]
@endreymarcell-wise thanks - any additional details would be greatly appreciated

I just tried the following: 

- `git clone https://github.com/twilio/twilio-cli`
- `npm install`
- `bin/dev config list` 

And didn't see the issue. Is there a particular command of yours that this is happening on?

I also verified that the 1.19.0 is being used throughout the dependency tree:

```
~/repos/twilio/twilio-cli (main*) » jq .version  node_modules/@oclif/core/package.json
"1.19.0"
```

**Comment by mdonnalley** [comment]
Thanks for the context - I was able to reproduce it with the following:

```
npm install -g twilio-cli
twilio plugins:install @twilio-labs/plugin-flex
twilio flex:plugins:create twilio-test
cd twilio-test
twilio flex:plugins:start
```

I'll report back once I know more about what's causing this

**Comment by shrutiburman** [comment]
@mdonnalley, steps to reproduce:
```
1. npm install -g twilio-cli
2. twilio plugins:install @twilio-labs/plugin-flex
3. flex plugins create <name_of_directory>
4. cd <name_of_directory>
5. npm install
6. flex plugins start
```

**Comment by mdonnalley** [comment]
@endreymarcell-wise The issue is that you're passing the `parse` method in [here](https://github.com/twilio/flex-plugin-builder/blob/main/packages/plugin-flex/src/sub-commands/flex-plugin.ts#L678) without binding `this` to it

Changing `super.parse` to `super.parse.bind(this)` should resolve the issue.

Not binding `this` worked for you up to this point because `parse` didn't access any class properties or methods until the change introduced in 1.19.0

**Comment by shrutiburman** [comment]
I'll get in touch with the respective team internally at Twilio. Thanks @mdonnalley !

**Comment by mdonnalley** [comment]
@endreymarcell-wise @shrutiburman happy to help. I'm going to close this issue for now but feel free to reopen it if you feel it's necessary

**Comment by shrutiburman** [comment]
Folks following this thread, flex has released a patch with the fix, [v6.0.2](https://github.com/twilio/flex-plugin-builder/commit/7eb1c2e7c728da1ee7475b6746f97816a1b6e9f7), that should solve this problem.
